x/rollapp/types/genesis_bridge_data_validator.go

```diff
@@ -0,0 +1,176 @@
+		channelID: rollappChannelID,
+		rollappID: rollappID,
+	}
+}
+
+type ValidationResult struct {
+	NativeDenomTrace  transfertypes.DenomTrace                // the ibc denom trace of the native rollapp denom
+	NativeDenom       banktypes.Metadata                      // the metadata of the native rollapp denom
+	GenesisAccPackets []transfertypes.FungibleTokenPacketData // packets holding native tokens dedicated for genesis accounts
+}
+
+func (v *GenesisBridgeValidator) Validate() (*ValidationResult, error) {
+	if err := v.rollapp.ValidateBasic(); err != nil {
+		return nil, errorsmod.Wrap(err, "validate basic genesis bridge data")
+	}
+
+	if err := validateAgainstHub(v.rollapp.GenesisInfo, v.hub); err != nil {
+		return nil, errorsmod.Wrap(err, "validate against rollapp")
+	}
+
+	trace, denomMeta, err := v.validateNativeDenom()
+	if err != nil {
+		return nil, errorsmod.Wrap(err, "validate native denom")
+	}
+
+	genesisAccPackets, err := v.validateGenesisTransfer()
+	if err != nil {
+		return nil, errorsmod.Wrap(err, "validate genesis transfer")
+	}
+
+	return &ValidationResult{
+		NativeDenomTrace:  trace,
+		NativeDenom:       denomMeta,
+		GenesisAccPackets: genesisAccPackets,
+	}, nil
+}
+
+func validateAgainstHub(packet GenesisBridgeInfo, hub GenesisInfo) error {
+	if packet.GenesisChecksum != hub.GenesisChecksum {
+		return fmt.Errorf("genesis checksum mismatch: expected: %v, got: %v", hub.GenesisChecksum, packet.GenesisChecksum)
+	}
+
+	if packet.Bech32Prefix != hub.Bech32Prefix {
+		return fmt.Errorf("bech32 prefix mismatch: expected: %v, got: %v", hub.Bech32Prefix, packet.Bech32Prefix)
+	}
+
+	if packet.NativeDenom != hub.NativeDenom {
+		return fmt.Errorf("native denom mismatch: expected: %v, got: %v", hub.NativeDenom, packet.NativeDenom)
+	}
+
+	if !packet.InitialSupply.Equal(hub.InitialSupply) {
+		return fmt.Errorf("initial supply mismatch: expected: %v, got: %v", hub.InitialSupply, packet.InitialSupply)
+	}
+
+	err := compareGenesisAccounts(hub.Accounts(), packet.GenesisAccounts)
+	if err != nil {
+		return errorsmod.Wrap(err, "genesis accounts mismatch")
+	}
+	return nil
+}
+
+func compareGenesisAccounts(raCommitted []GenesisAccount, gbData []GenesisAccount) error {
+	if len(raCommitted) != len(gbData) {
+		return fmt.Errorf("genesis accounts length mismatch: expected %d, got %d", len(raCommitted), len(gbData))
+	}
+
+	for _, acc := range raCommitted {
+		found := slices.ContainsFunc(gbData, func(dataAcc GenesisAccount) bool {
+			return dataAcc.Address == acc.Address && dataAcc.Amount.Equal(acc.Amount)
+		})
+
+		if !found {
+			return fmt.Errorf("genesis account mismatch: account %s with amount %v not found in data", acc.Address, acc.Amount)
+		}
+	}
+
+	return nil
+}
+
+// validateNativeDenom extracts the IBC denom trace and metadata from the rollapp native denom.
+// Resulting IBC denom is validated.
+func (v *GenesisBridgeValidator) validateNativeDenom() (transfertypes.DenomTrace, banktypes.Metadata, error) {
```

Comment:
this does alot more than just validating
I'd suggest to validate `v.rollapp.NativeDenom` here
but do the base change together with this:
```
	w.transferKeeper.SetDenomTrace(ctx, result.NativeDenomTrace)
	if err := w.denomKeeper.CreateDenomMetadata(ctx, result.NativeDenom); err != nil {
		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "create denom metadata"))
	}
```
after validation